src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducerBase.kt

```diff
@@ -8,98 +8,144 @@ package org.rust.cargo.runconfig.test
         sourceElement: Ref<PsiElement>
     ): Boolean {
-        val existingConfig = findConfig(context) ?: return false
-        sourceElement.set(existingConfig.sourceElement)
-        configuration.name = existingConfig.configurationName
-        val cmd = existingConfig.cargoCommandLine().mergeWithDefault(configuration)
+        val testConfig = findTestConfig(context) ?: return false
+        sourceElement.set(testConfig.sourceElement)
+        configuration.name = testConfig.configurationName
+        val cmd = testConfig.cargoCommandLine().mergeWithDefault(configuration)
         configuration.setFromCmd(cmd)
         return true
     }
 
-    fun findConfig(psi: Array<PsiElement>, climbUp: Boolean = true): TestConfig? {
-        if (psi.isEmpty()) return null
-        createMultipleFileConfiguration(psi.mapNotNull { findElement<RsMod>(it, climbUp) }.toTypedArray())
-            ?.let { return it }
-        findElement<RsFunction>(psi[0], climbUp)
-            ?.let { createSingleConfiguration(it) }
-            ?.let { return it }
-        return findElement<RsMod>(psi[0], climbUp)
-            ?.let { createSingleConfiguration(it) }
+    protected fun registerConfigProvider(provider: TestConfigProvider) {
+        testConfigProviders.add(provider)
     }
 
-    private fun findConfig(context: ConfigurationContext): TestConfig? {
-        val elements = LangDataKeys.PSI_ELEMENT_ARRAY.getData(context.dataContext)
-            ?: context.location?.psiElement?.let { arrayOf(it) }
-        return elements?.let { findConfig(it) }
+    fun findTestConfig(elements: List<PsiElement>, climbUp: Boolean = true): TestConfig? {
+        for (provider in testConfigProviders) {
+            val config = provider(elements, climbUp)
+            if (config != null) return config
+        }
+        return null
     }
 
-    private fun createSingleConfiguration(element: RsQualifiedNamedElement): TestConfig? {
-        if (!isSuitable(element)) return null
-        val configPath = element.crateRelativePath.configPath() ?: return null
-        val target = element.containingCargoTarget ?: return null
-        return TestConfig.SingleTestConfig(commandName, configPath, target, element)
+    private fun findTestConfig(context: ConfigurationContext): TestConfig? {
+        val elements = LangDataKeys.PSI_ELEMENT_ARRAY.getData(context.dataContext)?.toList()
+            ?: context.location?.psiElement?.let { listOf(it) }
+        return elements?.let { findTestConfig(it) }
     }
 
-    private fun createMultipleFileConfiguration(modules: Array<RsMod>): TestConfig? {
-        val modulesWithBenches = modules.filter { isSuitable(it) && it.containingCargoTarget != null }
+    private fun createConfigForDirectory(elements: List<PsiElement>): TestConfig? {
+        val dir = elements.singleOrNull() as? PsiDirectory ?: return null
+        val filesConfig = createConfigForMultipleFiles(dir.rustFiles, climbUp = false) ?: return null
```

Comment:
`rustFiles` is a very expensive operation because it reads the whole project
Can we avoid using it? It seems now you need these files only to find target(s). But looks like you can find the necessary target using only `dir` itself